PocketKit/Sources/RecentSavesWidgetKit/RecentSaves/Provider/RecentSavesProvider.swift

```diff
@@ -0,0 +1,166 @@
+    private func numberOfItems(for widgetFamily: WidgetFamily) -> Int {
+        switch widgetFamily {
+        case .systemExtraLarge:
+            return 0
+        case .systemLarge:
+            return 4
+        case .systemMedium:
+            return 2
+        default:
+            return 1
+        }
+    }
+
+    /// Builds an instance of `RecentSavesWidgetService`
+    /// - Returns: the instance
+    private func makeService() throws -> RecentSavesWidgetService {
+        let defaults = try makeUserDefaults()
+        return RecentSavesWidgetService(store: RecentSavesWidgetStore(userDefaults: defaults))
+    }
+
+    /// Returns the shared `UserDefaults` for the given build configuration
+    private func makeUserDefaults() throws -> UserDefaults {
+        guard let info = Bundle.main.infoDictionary,
+                let groupID = info["GroupId"] as? String,
+                let defaults = UserDefaults(suiteName: groupID) else {
+            throw RecentSavesProviderError.invalidStore
+        }
+        return defaults
+    }
+}
+
+// MARK: download thumbnails
+extension RecentSavesProvider {
+    /// Default size for downloaded thumbnails
+    static let defaultThumbnailSize = CGSize(width: 48, height: 36)
+
+    /// Download thumbnails, attach them to the related item and return the updated list of recent `[SavedItemRowContent]`
+    /// - Parameter content: the recent saves without thumbnails `[SavedItemContent]`
+    /// - Returns: the updated list
+    private func getContentWithImages(content: [SavedItemContent]) async -> [SavedItemRowContent] {
+        return await withTaskGroup(of: SavedItemRowContent.self, returning: [SavedItemRowContent].self) { taskGroup in
+            content.forEach { item in
+                    taskGroup.addTask {
+                        await downloadImage(for: item)
+                    }
+            }
+            // we need to update the content in the existing order,
+            // because we don't know when each task will complete.
+            let orderedContent = content.map { SavedItemRowContent(content: $0, image: nil) }
+            return await taskGroup.reduce(into: orderedContent) { orderedContent, item in
+                if let index = orderedContent.firstIndex(where: { $0.content == item.content }) {
+                    orderedContent[index] = item
+                }
+            }
+        }
+    }
+
+    /// Downloads the thumbnail for a given `SavedItemRowContent` item
+    /// - Parameter item: the given item
+    /// - Returns: the updated item with the downloaded image, or the original content, if no thumbnail was found.
+    private func downloadImage(for item: SavedItemContent) async -> SavedItemRowContent {
+        guard let imageUrl = item.imageUrl, let url = URL(string: imageUrl) else {
+            return SavedItemRowContent(content: item, image: nil)
+        }
+        do {
+            let (data, _) = try await URLSession.shared.data(from: bestURL(for: url))
```

Comment:
So I thought about using Kingfisher, but in reality, `URLSession` should still use its own cache, so, for the simple case of the thumbnails, I thought it could still work and avoid adding another dependency? That said, we can totally think of using KingFisher here